app/cdap/components/DataPrep/Directives/Calculate/index.js

```diff
@@ -466,51 +511,47 @@ export default class Calculate extends Component {
 
   parseColumns = () => {
     let columns = typeof this.props.column === 'string' ? [this.props.column] : this.props.column;
-
-    let type1 = DataPrepStore.getState().dataprep.typesCheck[columns[0]];
-    if (columns.length === 1) {
-      return {
-        isDisabled: false,
-        crossColumn: false,
-        columns: columns[0],
-        columnType: type1
-      };
-    }
-
-    if (columns.length === 2) {
-      const type2 = DataPrepStore.getState().dataprep.typesCheck[columns[1]];
-      let index = columns.indexOf(this.props.ddSelected);
-      if (type1 === type2 && index !== -1) {
-        columns = index === 0 ? columns : [this.props.ddSelected, columns[0]];
-        return {
-          isDisabled: false,
-          crossColumn: true,
-          columns: columns,
-          columnType: type1
-        };
+    let columnTypes = [...new Set(columns.map(col => DataPrepStore.getState().dataprep.typesCheck[col]))];
+    let crossColumn = columns.length === 2;
+    let multiColumn = columns.length >= 2;
+    let areColsMixedTypes = columnTypes.length > 1;
+    let allOptions;
+    if (multiColumn) {
+      if (crossColumn) {
+        allOptions = this.CROSS_COLUMN_CALCULATE_OPTIONS.concat(this.MULTI_COLUMN_CALCULATE_OPTIONS);
+      } else {
+        allOptions = this.MULTI_COLUMN_CALCULATE_OPTIONS;
       }
+    } else {
+      allOptions = this.CALCULATE_OPTIONS;
     }
+    let availableOptions = allOptions.filter(
+        option => option.name === 'label'
+                   || (!areColsMixedTypes && option.validColTypes.includes(columnTypes[0]))
+                   || (areColsMixedTypes
+                       && option.acceptMixedTypes
+                       && columnTypes.every(type => option.validColTypes.includes(type)))
+      );
+    let isDisabled = multiColumn && (!columns.includes(this.props.ddSelected) || availableOptions.length === 0);
 
-    type1 = DataPrepStore.getState().dataprep.typesCheck[columns];
     return {
-      isDisabled: true,
-      crossColumn: false,
-      columns: columns,
-      columnType: type1
+      isDisabled: isDisabled,
+      crossColumn: crossColumn,
+      multiColumn: multiColumn,
+      columns: multiColumn ? columns[0] : columns,
```

Comment:
if multicolumn use only the first column?

Reply:
Not sure why I had this -- removed, thanks!